File: go/chat_gpt/chat_gpt.go
package chat_gpt

import (
	"context"
	"encoding/json"
	"io/ioutil"
	"math/rand"
	"net/http"
	"strings"
	"time"
)

const saGPTHost = "https://chatgptproxy.xyz"

// ChatGPT 每一轮对话，都需要new一个新的
type ChatGPT struct {
	userFakeId string
	sessionId  string
	parentId   string
	stopHeart  bool
}

type ChatQuestionResp struct {
	Code     int    `json:"code"`
	CodeMsg  string `json:"code_msg"`
	TraceId  string `json:"trace_id"`
	RespData struct {
		ChatId string `json:"chat_id"`
	} `json:"resp_data"`
}

type ChatAnswerResp struct {
	Code     int    `json:"code"`
	CodeMsg  string `json:"code_msg"`
	TraceId  string `json:"trace_id"`
	RespData struct {
		Answer string `json:"answer"`
		Status int    `json:"status"`
	} `json:"resp_data"`
}

func NewChatGPT(userFakeId string) *ChatGPT {
	chat := &ChatGPT{
		userFakeId: userFakeId,
		sessionId:  randomString(16),
		parentId:   "0",
	}

	go chat.startHeart()

	return chat
}

func (c *ChatGPT) Close(ctx context.Context) {
	c.stopHeart = true
}

func (c *ChatGPT) SendQuestion(ctx context.Context, question string) error {
	apiPath := "/api/v1/chat/conversation"
	bodyData := map[string]interface{}{
		"data": map[string]string{
			"user_fake_id": c.userFakeId,
			"session_id":   c.sessionId,
			"parent_id":    c.parentId,
			"question":     question,
		},
	}
	dataByte, _ := json.Marshal(bodyData)
	reader := strings.NewReader(string(dataByte))
	resp, err := http.Post(saGPTHost+apiPath, "application/json", reader)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	respBody, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	qRespData := ChatQuestionResp{}
	if err := json.Unmarshal(respBody, &qRespData); err != nil {
		return err
	}

	c.parentId = qRespData.RespData.ChatId
	return nil
}

func (c *ChatGPT) GetAnswer(ctx context.Context, wait bool) (answer string, status int) {
	if !wait {
		return c.getAnswer(ctx)
	}
	for {
		answer, status = c.getAnswer(ctx)
		if status != 1 {
			return
		}
		time.Sleep(time.Second)
	}
}

func (c *ChatGPT) getAnswer(ctx context.Context) (answer string, status int) {
	apiPath := "/api/v1/chat/result"
	bodyData := map[string]interface{}{
		"data": map[string]string{
			"user_fake_id": c.userFakeId,
			"session_id":   c.sessionId,
			"chat_id":      c.parentId,
		},
	}
	dataByte, _ := json.Marshal(bodyData)
	reader := strings.NewReader(string(dataByte))
	resp, err := http.Post(saGPTHost+apiPath, "application/json", reader)
	if err != nil {
		return "", 4
	}
	defer resp.Body.Close()

	respBody, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", 4
	}

	respData := ChatAnswerResp{}
	if err := json.Unmarshal(respBody, &respData); err != nil {
		return "", 4
	}

	return respData.RespData.Answer, respData.RespData.Status
}

func (c *ChatGPT) startHeart() {
	for {
		if c.stopHeart {
			return
		}

		apiPath := "/api/v1/chat/heart"
		bodyData := map[string]interface{}{
			"data": map[string]string{
				"user_fake_id": c.userFakeId,
				"session_id":   c.sessionId,
			},
		}
		dataByte, _ := json.Marshal(bodyData)
		reader := strings.NewReader(string(dataByte))
		_, _ = http.Post(saGPTHost+apiPath, "application/json", reader)

		time.Sleep(time.Second * 10)
	}
}

func randomString(l int) string {
	bytes := []byte("ABCDEFGHJKMNPQRSTWXYZabcdefhijkmnprstwxyz2345678")

	result := make([]byte, l)
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	for i := 0; i < l; i++ {
		result[i] = bytes[r.Intn(len(bytes))]
	}
	return string(result)
}

File: python/chat_gpt.py
# -*- coding: utf-8 -*-

import time
import json
import string
import random
import threading

import requests


sa_chatGPT_host = "https://chatgptproxy.xyz"


class ChatGPT(object):
    def __init__(self, union_id, start_heart):
        self.user_fake_id = union_id
        self.session_id = random_str(16)
        self.parent_id = "0"

        self.heart = None
        if start_heart:
            self.heart = HealthThread(self.user_fake_id, self.session_id)
            self.heart.start()

    # reset session主要为了开启另一个对话，如果需要串成串的对话，不需要调用这个函数，使用一个session继续send_question即可
    def reset_session(self):
        if self.heart:
            self.heart.stop()
        self.session_id = random_str(16)
        self.parent_id = "0"

    # 关闭心跳
    def close(self):
        if self.heart:
            self.heart.stop()

    # 发送问题
    def send_question(self, question):
        api_path = "/api/v1/chat/conversation"
        data = {
            "data": {
                "user_fake_id": self.user_fake_id,
                "session_id": self.session_id,
                "parent_id": self.parent_id,
                "question": question
            }
        }
        resp = requests.post(
            url=sa_chatGPT_host + api_path,
            json=data
        )
        res_data = json.loads(resp.text)
        if 'resp_data' in res_data and 'chat_id' in res_data['resp_data']:
            self.parent_id = res_data['resp_data']['chat_id']
        else:
            print(res_data)
        return res_data['code'] == 200

    # 获取结果，wait标识，是否一只卡在这等待最终结果
    def get_result(self, wait):
        if not wait:
            return self._get_result()
        while True:
            answer, status = self._get_result()
            if status != 1:  # 1生成中，3完成，其他貌似都是错误 参见 js getChatResultWay()
                return answer, status
            time.sleep(1)

    def _get_result(self):
        api_path = "/api/v1/chat/result"
        data = {
            "data": {
                "user_fake_id": self.user_fake_id,
                "session_id": self.session_id,
                "chat_id": self.parent_id
            }
        }
        resp = requests.post(url=sa_chatGPT_host+api_path, json=data)
        res_data = json.loads(resp.text)
        # print(res_data)
        return res_data['resp_data']['answer'], res_data['resp_data']['status']


class HealthThread(threading.Thread):
    def __init__(self, user_fake_id, session_id):
        threading.Thread.__init__(self)
        self.user_fake_id = user_fake_id
        self.session_id = session_id
        self.stop_flag = False

    def run(self):
        while True:
            try:
                if self.stop_flag:
                    return
                time.sleep(10)
                self.heart()
            except Exception as exception:
                print(exception)

    def heart(self):
        api_path = "/api/v1/chat/heart"
        data = {
            "data": {
                "user_fake_id": self.user_fake_id,
                "session_id": self.session_id,
            }
        }
        requests.post(url=sa_chatGPT_host+api_path, json=data)

    def stop(self):
        self.stop_flag = True


def random_str(str_len):
    letters = string.ascii_lowercase
    return ''.join(random.choice(letters) for x in range(str_len))

